src/treeview_switches.ml

```diff
@@ -0,0 +1,207 @@
+    | Dependency (dep, _) -> Opam.Package.synopsis dep
+    | Switch (Named _) -> None
+    | Switch (Local _) -> None
+
+  let context_value = function
+    | Dependency (dep, _) -> (
+      match Opam.Package.documentation dep with
+      | Some _ -> "package-with-doc"
+      | None -> "package" )
+    | Switch _ -> "switch"
+
+  let icon ~extension_path = function
+    | Dependency _ ->
+      TreeItem.LightDarkIcon.
+        { light = `String (extension_path ^ "/assets/number-light.svg")
+        ; dark = `String (extension_path ^ "/assets/number-dark.svg")
+        }
+    | Switch _ ->
+      TreeItem.LightDarkIcon.
+        { light = `String (extension_path ^ "/assets/dependency-light.svg")
+        ; dark = `String (extension_path ^ "/assets/dependency-dark.svg")
+        }
+
+  let collapsible_state = function
+    | Dependency _ -> Vscode.TreeItemCollapsibleState.None
+    | Switch _ -> Vscode.TreeItemCollapsibleState.Collapsed
+end
+
+let make_item ~extension_path dependency =
+  let open Promise.Syntax in
+  let icon = `LightDark (Dependency.icon ~extension_path dependency) in
+  let collapsibleState = Dependency.collapsible_state dependency in
+  let label =
+    Vscode.TreeItemLabel.create ~label:(Dependency.label dependency) ()
+  in
+  let item = Vscode.TreeItem.make ~label ~collapsibleState () in
+  TreeItem.set_id item (Dependency.to_string dependency);
+  TreeItem.set_iconPath item icon;
+  TreeItem.set_contextValue item (Dependency.context_value dependency);
+  let+ _ =
+    Promise.Option.iter
+      (fun desc -> TreeItem.set_description item (`String desc))
+      (Dependency.description dependency)
+  in
+  Option.iter (Dependency.tooltip dependency) ~f:(fun desc ->
+      TreeItem.set_tooltip item (`String desc));
+  item
+
+let get_dependency_dependencies _dependency = Promise.return (Some [])
+
+let get_switch_dependencies switch =
+  let open Promise.Syntax in
+  let* packages = Opam.get_switch_packages switch in
+  match packages with
+  | Ok packages ->
+    let names =
+      packages
+      |> List.map ~f:(fun n ->
+             Dependency.Dependency (n, Some (Dependency.Switch switch)))
+    in
+    Promise.return (Some names)
+  | Error err ->
+    let* _ =
+      Vscode.Window.showInformationMessage
+        ~message:
+          (Printf.sprintf
+             "An error occured while reading the switch dependencies: %s" err)
+        ()
+    in
+    Promise.return None
+
+let get_dependencies ~extension_path element =
+  let open Promise.Syntax in
+  let dependency =
+    match Vscode.TreeItem.id element with
+    | None ->
+      (* This should never happen, the elements are always created with an ID. *)
+      assert false
+    | Some dep -> Dependency.of_string dep
+  in
+  let dependencies_opt_to_tree_items deps_opt =
+    let* deps_opt = deps_opt in
+    match deps_opt with
+    | None -> Promise.return None
+    | Some deps ->
+      let* items =
+        List.fold_left ~init:(Promise.return [])
```

Comment:
I'd add `Promise.List.all` to `promise_jsoo`. It's ok to just leave a TODO for now.

Reply:
Pushed here: https://github.com/mnxn/promise_jsoo/pull/5

Reply:
So this can just use `Promise.all_list` for now.